File: src/RecordProcessor.ts
import {Writable} from "stream"
import {ISourceMeta, PkMap, PKType} from "./jsonSchemaInspector"
import {extractValue} from "./jsonSchemaTranslator"
import {log_debug, log_info} from "singer-node"
import TargetConnection from "./TargetConnection"

// eslint-disable-next-line @typescript-eslint/no-var-requires
const get = require("lodash.get")

type PKValue = string | number
type PKValues = PKValue[]

type SourceMetaPK = {
  rootValues: PKValues | undefined,
  parentValues: PKValues | undefined,
  values: PKValues,
  levelValues: PKValues | undefined,
};

interface RecordProcessorConfig {
  batchSize: number
  translateValues: boolean
  autoEndTimeoutMs: number
}

interface IngestionCtx {
  stream: Writable
  promise: Promise<void>
  autoEndTimeout: NodeJS.Timeout
}

/**
 * Ingests and store data values
 * Tree structure to process stream of data according to precomputed meta data
 * Call pushRecord for each row
 * One node for one table
 */
export default class RecordProcessor {
  readonly hasChildren: boolean
  private readonly isWithParentPK: boolean
  private readonly isRoot: boolean
  private readonly children: { [key: string]: RecordProcessor }
  private ingestionCtx?: IngestionCtx
  private bufferedDatasToStream: string[] = []
  private readonly currentPkMappings: PkMap[]

  constructor(
    private readonly meta: ISourceMeta,
    private readonly clickhouse: TargetConnection,
    private readonly config: RecordProcessorConfig,
    private readonly level = 0,
  ) {
    this.meta = meta
    this.isRoot = level === 0
    this.isWithParentPK = !this.isRoot && this.meta.pkMappings.find((pk) => pk.pkType === PKType.PARENT) !== undefined
    this.hasChildren = meta.children.length > 0
    this.children = meta.children.reduce((acc, child) => {
      const processor = new RecordProcessor(child, this.clickhouse, this.config, this.level + 1)
      return {...acc, [child.sqlTableName]: processor}
    }, {})
    this.currentPkMappings = this.meta.pkMappings.filter((pkMap) => pkMap.pkType === PKType.CURRENT)
  }

  /*
      Prepare sql query by splitting fields and values
      Creates children
      Structure: One child per table
   */
  pushRecord(
    data: Record<string, any>,
    abort: (err: Error) => void,
    maxVer: number,
    parentMeta?: SourceMetaPK,
    rootVer?: number,
    indexInParent = -1,
    messageCount = 0,
  ): void {
    // When first record is pushed we start by initializing stream
    if (!this.isInitialized()) {
      this.startIngestion(messageCount, abort)
    }
    this.ingestionCtx!.autoEndTimeout.refresh()

    // root version number is computed only for a root who has primaryKeys
    // version start at max existing version + 1
    const resolvedRootVer = (this.isRoot && this.meta.pkMappings.length > 0) ? maxVer + 1 : rootVer

    const currentPkValues = new Array(this.currentPkMappings.length)
    for (let i = 0; i < this.currentPkMappings.length; i++) {
      currentPkValues[i] = extractValue(data, this.currentPkMappings[i], this.config.translateValues)
    }

    const sourceMetaPK: SourceMetaPK = {
      values: currentPkValues,
      rootValues: this.isRoot ? undefined : (parentMeta?.rootValues ?? parentMeta?.values),
      parentValues: this.isRoot ? undefined : parentMeta?.values,
      levelValues: this.isRoot ? undefined : [...(parentMeta?.levelValues || []), indexInParent],
    }

    let pkValues = currentPkValues
    if (!this.isRoot) {
      pkValues = sourceMetaPK.rootValues!
        .concat(this.isWithParentPK ? sourceMetaPK.parentValues! : [])
        .concat(pkValues)
        .concat(sourceMetaPK.levelValues!)
    }

    const dataToStream = JSON.stringify(this.buildSQLInsertValues(data, pkValues, resolvedRootVer))
    this.bufferedDatasToStream.push(dataToStream)
    if (this.bufferedDatasToStream.length == this.config.batchSize) {
      this.sendBufferedDatasToStream()
    }

    if (this.hasChildren) {
      for (const child of this.meta.children) {
        const childProcessor: RecordProcessor = this.children[child.sqlTableName]
        // In this record we expect a list, as that is the reason a children has been created. But some JSON Schema declaration may declare both an array and a list, so we check and create an array with only one item if it is not an array
        const childDataRaw: Record<string, any> = get(data, child.prop.split("."))
        const childDataAsArray = Array.isArray(childDataRaw) ? childDataRaw : (childDataRaw ? [childDataRaw] : [])
        for (let idx = 0; idx < childDataAsArray.length; idx++) {
          childProcessor.pushRecord(childDataAsArray[idx], abort, maxVer, sourceMetaPK, resolvedRootVer, idx, messageCount)
        }
      }
    }
  }

  public async endIngestion() {
    if (this.isInitialized()) {
      log_debug(`closing stream to insert data in ${this.meta.prop}, ${this.meta.sqlTableName}`)
      const {promise, stream, autoEndTimeout} = this.ingestionCtx!
      clearTimeout(autoEndTimeout)
      this.sendBufferedDatasToStream()
      stream.end()
      this.ingestionCtx = undefined
      await Promise.all([
        promise,
        Promise.all(Object.values(this.children).map((child) => child.endIngestion())),
      ])
    }
  }

  public buildSQLInsertField(): string[] {
    const isRoot = this.meta.pkMappings.find((pkMap) => pkMap.pkType === PKType.ROOT) === undefined
    return this.meta.pkMappings
      .map((pkMap) => pkMap.sqlIdentifier)
      .concat(this.meta.simpleColumnMappings.map((cMap) => cMap.sqlIdentifier))
      .concat(isRoot ? (this.meta.pkMappings.length > 0 ? ["`_ver`"] : []) : ["`_root_ver`"])
  }

  private isInitialized(): boolean {
    return this.ingestionCtx !== undefined
  }

  private sendBufferedDatasToStream() {
    if (this.bufferedDatasToStream.length > 0) {
      this.bufferedDatasToStream.push("")
      if (!this.ingestionCtx) {
        throw new Error("ingestionCtx is undefined but there is still bufferedData")
      }
      this.ingestionCtx.stream.write(Buffer.from(this.bufferedDatasToStream.join('\n')))
      this.bufferedDatasToStream = []
    }
  }

  private startIngestion(messageCount: number, abort: (err: Error) => void): void {
    const insertQuery = `INSERT INTO ${this.meta.sqlTableName} (${this.buildSQLInsertField().join(",")}) FORMAT JSONCompactEachRow`
    if (this.isRoot) {
      log_info(`[${this.meta.prop}] handling lines starting at ${messageCount}`)
    }

    let promiseResolve: (value: (void | PromiseLike<void>)) => void
    let promiseReject: (reason?: any) => void;

    const promise = new Promise<void>((resolve, reject) => {
      promiseResolve = resolve;
      promiseReject = reject;
    })

    this.ingestionCtx = {
      stream: this.clickhouse.createWriteStream(insertQuery, (err: any) => {
        if (err) {
          abort(err)
          promiseReject(err)
        } else {
          promiseResolve()
        }
      }),
      promise,
      autoEndTimeout: setTimeout(() => {
        log_debug(`auto closing stream to insert data in ${this.meta.prop}, ${this.meta.sqlTableName} due to inactivity`)
        this.endIngestion()
      }, this.config.autoEndTimeoutMs),
    }
  }

  // Extract value for all simple columns in a record
  private buildSQLInsertValues = (
    data: Record<string, any>,
    pkValues: any[] = [],
    version?: number,
  ) => {
    const noPk = pkValues.length
    const noSimpleColumn = this.meta.simpleColumnMappings.length
    const result: any[] = new Array(noPk + noSimpleColumn + (version !== undefined ? 1 : 0))
    for (let i = 0; i < noPk; i++) {
      result[i] = pkValues[i]
    }
    for (let i = 0; i < noSimpleColumn; i++) {
      result[i + noPk] = extractValue(data, this.meta.simpleColumnMappings[i], this.config.translateValues)
    }
    if (version !== undefined)
      result[noPk + noSimpleColumn] = version
    return result
  }
}

File: src/main/kotlin/com/biron/singerTargetClickhouse/RecordProcessor.kt
package com.biron.singerTargetClickhouse

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.jsonMapper
import com.fasterxml.jackson.module.kotlin.kotlinModule
import io.github.oshai.kotlinlogging.KotlinLogging
import java.io.ByteArrayOutputStream
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit

private val logger = KotlinLogging.logger {}

private val jsonMapper: ObjectMapper = jsonMapper { addModule(kotlinModule()) }

data class RecordProcessorConfig(
	val batchSize: Int,
	val translateValues: Boolean,
	val autoEndTimeoutMs: Long,
)

/**
 * Ingests and stores data values. Tree structure mirrors the precomputed source meta:
 * one node per table. Each node keeps its own insert stream open until the batch size
 * is reached or the auto-end timeout fires, then flushes and reopens on the next push.
 *
 * All mutations of [buffered]/[ingestion] take [lock], so the scheduler-thread-driven
 * auto-end cannot race with the caller thread.
 *
 * Rows arrive as [RecordRow]s — slot layout `[currentPks..., simpleColumns..., subtables...]`
 * determined by the matching [StreamReader]. The processor reads its data from the row by
 * slot index; translation (config.translateValues) already happened inside the reader.
 */
class RecordProcessor(
	private val meta: SourceMeta,
	private val clickhouse: TargetConnection,
	private val config: RecordProcessorConfig,
	private val level: Int = 0,
) {
	private val isRoot: Boolean = level == 0
	private val isWithParentPK: Boolean = !isRoot && meta.pkMappings.any { it.pkType == PKType.PARENT }
	val hasChildren: Boolean = meta.children.isNotEmpty()
	private val currentPkMappings: List<PkMap> = meta.pkMappings.filter { it.pkType == PKType.CURRENT }
	private val pkCount: Int = currentPkMappings.size
	private val columnCount: Int = meta.simpleColumnMappings.size
	private val children: Map<String, RecordProcessor> = meta.children
		.associateBy { it.sqlTableName }
		.mapValues { RecordProcessor(it.value, clickhouse, config, level + 1) }

	private val lock = Any()
	private var ingestion: Ingestion? = null
	private val buffered: MutableList<List<Any?>> = mutableListOf()

	data class SourceMetaPK internal constructor(
		val values: List<Any?>,
		val rootValues: List<Any?>?,
		val parentValues: List<Any?>?,
		val levelValues: List<Any?>?,
	)

	private class Ingestion(
		val writer: RowWriter,
		var timeoutFuture: ScheduledFuture<*>?,
	)

	fun pushRecord(
		row: RecordRow,
		abort: (Throwable) -> Unit,
		maxVer: Long,
		parentMeta: SourceMetaPK? = null,
		rootVer: Long? = null,
		indexInParent: Int = -1,
		messageCount: Int = 0,
	) {
		val (sourcePk, resolvedRootVer) = synchronized(lock) {
			if (ingestion == null) startIngestion(messageCount, abort)

			val rootVersion = if (isRoot && meta.pkMappings.isNotEmpty()) maxVer else rootVer
			val srcPk = buildSourcePk(row, parentMeta, indexInParent)
			val pkValues = composePkValues(srcPk)

			buffered.add(buildInsertValues(row, pkValues, rootVersion))
			if (buffered.size >= config.batchSize) {
				refreshTimeout(abort)
				flushBuffered()
			}

			srcPk to rootVersion
		}

		if (hasChildren) dispatchToChildren(row, sourcePk, resolvedRootVer, maxVer, messageCount, abort)
	}

	private fun buildSourcePk(row: RecordRow, parentMeta: SourceMetaPK?, indexInParent: Int): SourceMetaPK {
		val currentPkValues: List<Any?> = List(pkCount) { row[it] }
		return SourceMetaPK(
			values = currentPkValues,
			rootValues = if (isRoot) null else parentMeta?.rootValues ?: parentMeta?.values,
			parentValues = if (isRoot) null else parentMeta?.values,
			levelValues = if (isRoot) null else (parentMeta?.levelValues ?: emptyList()) + indexInParent,
		)
	}

	private fun composePkValues(srcPk: SourceMetaPK): List<Any?> {
		if (isRoot) return srcPk.values
		return buildList {
			addAll(srcPk.rootValues!!)
			if (isWithParentPK) addAll(srcPk.parentValues!!)
			addAll(srcPk.values)
			addAll(srcPk.levelValues!!)
		}
	}

	private fun dispatchToChildren(
		row: RecordRow,
		sourcePk: SourceMetaPK,
		rootVer: Long?,
		maxVer: Long,
		messageCount: Int,
		abort: (Throwable) -> Unit,
	) {
		meta.children.forEachIndexed { i, child ->
			val processor = children.getValue(child.sqlTableName)
			@Suppress("UNCHECKED_CAST")
			val childRows = row[pkCount + columnCount + i] as? List<RecordRow> ?: return@forEachIndexed
			childRows.forEachIndexed { idx, childRow ->
				processor.pushRecord(
					row = childRow,
					abort = abort,
					maxVer = maxVer,
					parentMeta = sourcePk,
					rootVer = rootVer,
					indexInParent = idx,
					messageCount = messageCount,
				)
			}
		}
	}

	fun endIngestion() {
		synchronized(lock) {
			val ctx = ingestion ?: return@synchronized
			logger.debug { "[${meta.prop}] closing stream to insert data" }
			ctx.timeoutFuture?.cancel(false)
			flushBuffered()
			try {
				ctx.writer.close()
			} finally {
				ingestion = null
			}
		}
		children.values.forEach { it.endIngestion() }
	}

	internal fun buildSQLInsertField(): List<String> {
		val noRootPk = meta.pkMappings.none { it.pkType == PKType.ROOT }
		return meta.pkMappings.map { it.sqlIdentifier } +
			meta.simpleColumnMappings.map { it.sqlIdentifier } +
			when {
				noRootPk && meta.pkMappings.isNotEmpty() -> listOf("`_ver`")
				noRootPk -> emptyList()
				else -> listOf("`_root_ver`")
			}
	}

	private fun flushBuffered() {
		if (buffered.isEmpty()) return
		val ctx = ingestion ?: error("ingestion not started but buffered data present")
		// Write rows through a UTF-8 JsonGenerator directly into a byte buffer — avoids the
		// per-row StringWriter+StringBuilder round-trip and the final String.getBytes encode
		// that together dominated flushBuffered in the profile.
		val baos = ByteArrayOutputStream(8 * 1024)
		jsonMapper.factory.createGenerator(baos).use { gen ->
			// Jackson's default inserts " " between successive root-level values; disable that
			// so consecutive rows are separated only by the '\n' we write below.
			gen.setRootValueSeparator(null)
			buffered.forEach { row ->
				jsonMapper.writeValue(gen, row)
				gen.writeRaw('\n')
			}
		}
		ctx.writer.write(baos.toByteArray())
		buffered.clear()
	}

	private fun startIngestion(messageCount: Int, abort: (Throwable) -> Unit) {
		val insertQuery = "INSERT INTO ${meta.sqlTableName} (${buildSQLInsertField().joinToString(",")}) FORMAT JSONCompactEachRow"
		if (isRoot) logger.info { "[${meta.prop}] handling lines starting at $messageCount" }
		val writer = clickhouse.openRowWriter(insertQuery)
		val ctx = Ingestion(writer, timeoutFuture = null)
		ctx.timeoutFuture = scheduleAutoEnd(ctx, abort)
		ingestion = ctx
	}

	private fun refreshTimeout(abort: (Throwable) -> Unit) {
		val ctx = ingestion ?: return
		ctx.timeoutFuture?.cancel(false)
		ctx.timeoutFuture = scheduleAutoEnd(ctx, abort)
	}

	/**
	 * Schedule auto-end for [ctx]. The callback itself re-acquires [lock] and re-checks that
	 * the ingestion is still current: if another push replaced the timer (or closed the
	 * ingestion) between scheduling and firing, the callback becomes a no-op.
	 */
	private fun scheduleAutoEnd(ctx: Ingestion, abort: (Throwable) -> Unit): ScheduledFuture<*> {
		val taskRef = java.util.concurrent.atomic.AtomicReference<ScheduledFuture<*>?>(null)
		val task = defaultScheduler.schedule(
			{
				synchronized(lock) {
					if (ingestion !== ctx || ctx.timeoutFuture !== taskRef.get()) return@synchronized
					logger.debug { "[${meta.prop}] auto closing stream to insert data due to inactivity" }
					try {
						endIngestionUnlocked()
					} catch (e: Throwable) {
						abort(e)
					}
				}
			},
			config.autoEndTimeoutMs,
			TimeUnit.MILLISECONDS,
		)
		taskRef.set(task)
		return task
	}

	/** Close the current ingestion assuming [lock] is already held (or we are on the lock-thread). */
	private fun endIngestionUnlocked() {
		val ctx = ingestion ?: return
		flushBuffered()
		try {
			ctx.writer.close()
		} finally {
			ingestion = null
		}
	}

	private fun buildInsertValues(
		row: RecordRow,
		pkValues: List<Any?>,
		version: Long?,
	): List<Any?> = buildList(pkValues.size + columnCount + (if (version != null) 1 else 0)) {
		addAll(pkValues)
		for (i in 0 until columnCount) add(row[pkCount + i])
		if (version != null) add(version)
	}

	companion object {
		private val defaultScheduler: ScheduledExecutorService =
			Executors.newSingleThreadScheduledExecutor { r ->
				Thread(r, "record-processor-auto-end").apply { isDaemon = true }
			}
	}
}
